File: mentoring.cpp
// 1. Одиночка (Singleton) - C++
#include <iostream> using namespace std;

class Singleton { private: static Singleton* instance; Singleton() {}

public: static Singleton* getInstance() { if (!instance) instance = new Singleton(); return instance; } };

Singleton* Singleton::instance = nullptr;

int main() { Singleton* inst1 = Singleton::getInstance(); Singleton* inst2 = Singleton::getInstance(); cout << (inst1 == inst2) << endl; // true return 0; }

// 2. Фабричный метод (Factory Method) - C++ #include <iostream> using namespace std;

class Transport { public: virtual string deliver() = 0; };

class Truck : public Transport { public: string deliver() override { return "Delivering by land in a truck"; } };

class Ship : public Transport { public: string deliver() override { return "Delivering by sea in a ship"; } };

class TransportFactory { public: virtual Transport* createTransport() = 0; };

class TruckFactory : public TransportFactory { public: Transport* createTransport() override { return new Truck(); } };

class ShipFactory : public TransportFactory { public: Transport* createTransport() override { return new Ship(); } };

int main() { TransportFactory* factories[] = {new TruckFactory(), new ShipFactory()}; for (TransportFactory* factory : factories) { Transport* transport = factory->createTransport(); cout << transport->deliver() << endl; delete transport; } return 0; }

// 3. Строитель (Builder) - C++ #include <iostream> using namespace std;

class Car { public: string engine; int wheels; bool sunroof;

void show() {
    cout << "Car(engine=" << engine << ", wheels=" << wheels << ", sunroof=" << sunroof << ")" << endl;
}

};

class CarBuilder { private: Car car; public: CarBuilder& setEngine(string e) { car.engine = e; return *this; } CarBuilder& setWheels(int w) { car.wheels = w; return *this; } CarBuilder& setSunroof(bool s) { car.sunroof = s; return *this; } Car build() { return car; } };

int main() { Car car = CarBuilder().setEngine("V8").setWheels(4).setSunroof(true).build(); car.show(); return 0; }

File: mentoring.py
#1. Одиночка (Singleton) - Гарантирует, что класс имеет только один экземпляр.

class Singleton: _instance = None

def __new__(cls):
    if cls._instance is None:
        cls._instance = super(Singleton, cls).__new__(cls)
    return cls._instance

inst1 = Singleton() inst2 = Singleton() print(inst1 is inst2)  # True, так как оба указывают на один объект

#2. Фабричный метод (Factory Method) - Делегирует создание объектов подклассам.

from abc import ABC, abstractmethod

class Transport(ABC): @abstractmethod def deliver(self): pass

class Truck(Transport): def deliver(self): return "Delivering by land in a truck"

class Ship(Transport): def deliver(self): return "Delivering by sea in a ship"

class TransportFactory(ABC): @abstractmethod def create_transport(self): pass

class TruckFactory(TransportFactory): def create_transport(self): return Truck()

class ShipFactory(TransportFactory): def create_transport(self): return Ship()

factories = [TruckFactory(), ShipFactory()] for factory in factories: transport = factory.create_transport() print(transport.deliver())

#3. Строитель (Builder) - Упрощает создание сложных объектов

class Car: def init(self, engine=None, wheels=None, sunroof=None): self.engine = engine self.wheels = wheels self.sunroof = sunroof

def __repr__(self):
    return f"Car(engine={self.engine}, wheels={self.wheels}, sunroof={self.sunroof})"

class CarBuilder: def init(self): self.car = Car()

def set_engine(self, engine):
    self.car.engine = engine
    return self

def set_wheels(self, wheels):
    self.car.wheels = wheels
    return self

def set_sunroof(self, sunroof):
    self.car.sunroof = sunroof
    return self

def build(self):
    return self.car

car = CarBuilder().set_engine("V8").set_wheels(4).set_sunroof(True).build() print(f"Built car: {car}")

